java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightStream.java

```diff
@@ -131,33 +131,48 @@ public DictionaryProvider takeDictionaryOwnership() {
     return provider;
   }
 
+  /**
+   * Get the descriptor for this stream. Only applicable on the server side of a DoPut operation. Will block until the
+   * client sends the descriptor.
+   */
   public FlightDescriptor getDescriptor() {
+    // This blocks until the schema message (with the descriptor) is sent.
+    getRoot();
     return descriptor;
   }
 
   /**
    * Closes the stream (freeing any existing resources).
    *
-   * <p>If the stream is isn't complete and is cancellable this method will cancel the stream first.</p>
+   * <p>If the stream isn't complete and is cancellable, this method will cancel the stream first.</p>
    */
   public void close() throws Exception {
-    if (!completed && cancellable != null) {
-      cancel("Stream closed before end.", null);
+    RuntimeException cancelException = null;
+    try {
+      // cancel() can throw, but we still want to clean up resources.
+      if (!completed && cancellable != null) {
+        cancel("Stream closed before end.", /* no exception to report */ null);
+      }
+    } catch (RuntimeException e) {
+      cancelException = e;
+    }
+
+    try {
+      final List<AutoCloseable> closeables = new ArrayList<>();
+      closeables.add(root.get());
+      closeables.add(applicationMetadata);
+      closeables.addAll(queue);
+      if (dictionaries != null) {
```

Comment:
For future reference you can also use Optional here if you like the fluent style.

Reply:
Yes, I'm a little ambivalent about using Optional in that style...it feels a little wasteful, even though I know it won't actually matter in practice.